fix CoordSystem._inverse_transformation method

alexcockburn1 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #21356 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The `CoordSystem._inverse_transformation` method was previously
throwing a ValueError for any input.  This change fixes the broken logic
and adds logic to raise an error if we cannot find an inverse for the
coordinate transformation.

The previous implementation converted variables to Dummy variables
so that all variables in the `solve` method have the same type, but in
this implementation I have converted the variables to 
CoordinateSymbols instead.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* diffgeom
  * Fixed a bug in the `CoordSystem._inverse_transformation` method.
<!-- END RELEASE NOTES -->

alexcockburn1 committed:
fix CoordSystem._inverse_transformation method

The CoordSystem._inverse_transformation method was previously
throwing a ValueError for any input.  This change fixes the broken logic
and adds logic to raise an error if we cannot find an inverse for the
coordinate transformation.

Halfogerty commented:
Is there a way to get this reviewed by someone?  This is my first contribution to any open source project, so please let me know if there's something more I should be doing to get this merged.

oscarbenjamin commented:
Thanks for contributing this. It sometimes takes a while for anyone to find time to review a pull request.

Personally I don't know the diffgeom module at all so it would be difficult for me to review this. I'll wait and see if anyone else can review it.

JSS95 commented:
I like this fix, but `CoordSystem` needs to be enhanced first. Current design with `Lambda` can be improved by storing the generator `Symbol`s to `CoordSystem`. Then we may not need transformations such as `to_coord_symbol()`.

Halfogerty commented:
Thanks for taking a look at this.  Could you clarify what you mean by "storing the generator `Symbol`s to `CoordSystem`"?

The way things work at the moment is that a relation_dict is needed to create a `CoordSystem`, and a `CoordSystem` is needed to create a `CoordinateSymbol`.  Therefore using `CoordinateSymbol` in the relation_dict would introduce a cyclic dependency, so we have to use `Symbol` in the relation_dict, and convert between `Symbol` and `CoordinateSymbol`.

Another option for avoiding `to_coord_symbol` would be to add a method to the `CoordSystem` class which returns the symbols as instances of `Symbol`,  in addition to the method which returns the symbols as instances of `CoordinateSymbol`, and only work with `Symbol`s in the `_inverse_tranformation` method.

JSS95 commented:
> Thanks for taking a look at this. Could you clarify what you mean by "storing the generator `Symbol`s to `CoordSystem`"?

It seems that I was confused. `CoordSystem` already stores the symbols. I opened a PR #21464 which makes the signature of `CoordSystem` simpler.

JSS95 commented:
@Halfogerty Will you pull and merge the master branch into this branch, please? Signature for `CoordSystem` is changed a little bit.

alexcockburn1 committed:
Merge branch 'master' into diff_geom_improvements

alexcockburn1 committed:
fix CoordSystem._inverse_transformation method

The relation_dict used to create CoordSystem has
changed upstream to use tuples rather than Lambdas.
This is a refactor to allow for that.

alexcockburn1 committed:
fix CoordSystem._inverse_transformation method

Remove unused Lambda import

Halfogerty commented:
@JSS95 Sure, I've updated the pull request now.

JSS95 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -398,26 +398,28 @@ def transformation(self, sys):
-        forward_transform = sys1.transform(sys2)
-        forward_syms, forward_results = forward_transform.args
+        forward_transform_expressions = sys1.transform(sys2)
+        # Coordinate transformations are defined in terms of Symbols rather than CoordSymbols, so must be converted
+        forward_transform_expressions = to_coord_symbol(sys1.symbols, forward_transform_expressions)
```
`sys1.transform(sys2)` already consists of `CoordinateSymbol` so we don't need `to_coord_symbol()` function. (Removing this line showed no different result)

Halfogerty replied:
Ah thanks, I hadn't realised that the `transform` method was doing this for me

JSS95 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -398,26 +398,28 @@ def transformation(self, sys):
+            list(sys1.symbols), dict=True)
+        if len(inv_results) == 0:
+            raise NotImplementedError(
+                "Cannot solve inverse transformation of {}".format(forward_transform_expressions))
```
I think the error message should tell which systems resulted the failure, e.g. `"Cannot solve inverse transformation between {} and {}".format(sys1, sys2)`.

Halfogerty replied:
Done

JSS95 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -398,26 +398,28 @@ def transformation(self, sys):
-        inv_results = [inv_results[s] for s in forward_syms]
+            [t[0] - t[1] for t in zip(sys2.symbols, forward_transform_expressions)],
+            list(sys1.symbols), dict=True)
+        if len(inv_results) == 0:
+            raise NotImplementedError(
+                "Cannot solve inverse transformation of {}".format(forward_transform_expressions))
```
An error should be raised for ambiguous result (i.e. `len(inv_results) > 1`) as well.

Halfogerty replied:
Done

JSS95 commented on sympy/diffgeom/tests/test_diffgeom.py:
```diff
@@ -36,6 +36,34 @@ def test_R2():
             R2_r, R2_r.coord_tuple_transform_to(R2_p, m)).applyfunc(simplify)
 
+
+def test_inverse_transformations():
+    p, q, r, s = symbols('p q r s')
+
+    relations_quarter_rotation = {
+        ('first', 'second'): (q, - p)
+    }
+
+    R2_pq = CoordSystem('first', R2_origin, [p, q], relations_quarter_rotation)
+    R2_rs = CoordSystem('second', R2_origin, [r, s], relations_quarter_rotation)
+
+    # The transform method should derive the inverse transformation if not given explicitly
+    assert R2_rs.transform(R2_pq) == Matrix([[- R2_rs.symbols[1]], [R2_rs.symbols[0]]])
+
+    a, b = symbols('a b', positive=True)
+    relations_uninvertible_transformation = {
+        ('first', 'second'): (- a,)
+    }
+
+    R2_a = CoordSystem('first', R2_origin, [a], relations_uninvertible_transformation)
+    R2_b = CoordSystem('second', R2_origin, [b], relations_uninvertible_transformation)
+
+    # The transform method should throw if it cannot invert the coordinate transformation.
+    # This transformation is uninvertible because there is no positive a, b satisfying a = -b
+    with raises(NotImplementedError):
+        R2_b.transform(R2_a)
```
Please remove the space between the negative sign and the expression (`- a` -> `-a`)

Halfogerty replied:
Done

alexcockburn1 committed:
fix CoordSystem._inverse_transformation method

Responding to reviewer comments:
* Removed to_coord_symbol function
* Improved error message for when we can't find the inverse of a
coordinate transformation
* Added an error for when we find more than one inverse to a
coordinate transformation
* Removed spaces after negative signs

JSS95 commented:
Looks good!
